src/js/plugins/plugin.chapters.js

```diff
@@ -0,0 +1,190 @@
+};
+
+/*
+ * Remove all chapters.
+ */
+BookReader.prototype.removeChapters = function() {
+    this.$('.BRnavpos .BRchapter').remove();
+};
+
+/**
+ * Update the table of contents based on array of TOC entries.
+ * @param { list } tocEntries
+ */
+BookReader.prototype.updateTOC = function(tocEntries) {
+    this.removeChapters();
+    for (let i = 0; i < tocEntries.length; i++) {
+        this.addChapterFromEntry(tocEntries[i]);
+    }
+};
+
+/**
+ *   Example table of contents entry - this format is defined by Open Library
+ *   {
+ *       "pagenum": "17",
+ *       "level": 1,
+ *       "label": "CHAPTER I",
+ *       "type": {"key": "/type/toc_item"},
+ *       "title": "THE COUNTRY AND THE MISSION"
+ *   }
+ * @param { object } - tocEntryObject
+ * @param { string } pageNum
+ * @param { string } title
+ */
+BookReader.prototype.addChapterFromEntry = function(tocEntryObject) {
+    const pageIndex = this.getPageIndex(tocEntryObject['pagenum']);
+    // Only add if we know where it is
+    if (pageIndex) {
+        this.addChapter(tocEntryObject['title'], tocEntryObject['pagenum'], pageIndex);
+    }
+    this.$('.BRchapter').each(function(){
+        $(this).hover(function(){
+            $(this).addClass('front');
+        },function(){
+            $(this).removeClass('front');
+        });
+    });
+    this.$('.BRsearch').each(function(){
+        $(this).hover(function(){
+            $(this).addClass('front');
+        },function(){
+            $(this).removeClass('front');
+        });
+    });
+};
+
+// getOpenLibraryRecord
+//
+// The bookreader is designed to call openlibrary API and constructs the
+// "Return book" button using the response.
+//
+// This makes a call to OL API and calls the given callback function with the
+// response from the API.
+BookReader.prototype.getOpenLibraryRecord = function () {
+  // Try looking up by ocaid first, then by source_record
+  const baseURL = `${this.olHost}//query.json?type=/type/edition&*=`;
+  const fetchUrlByBookId = `${baseURL}&ocaid=${this.bookId}`;
+
+  /*
+  * Update Chapter markers based on received record from Open Library.
+  * Notes that Open Library record is used for extra metadata, and also for lending
+  */
+  const setUpChapterMarkers = (olObject) => {
+    if (olObject && olObject.table_of_contents) {
+      // XXX check here that TOC is valid
+      this.updateTOC(olObject.table_of_contents);
+    }
+  };
+
+  $.ajax({
+    url: fetchUrlByBookId,
+    success: (data) => {
+      if (data && data.length > 0) {
+        setUpChapterMarkers(data[0]);
+      } else {
+        // try sourceid
+        const fetchURLBySourceId = `${baseURL}&source_records=ia:${this.bookId}`;
+        $.ajax({
+          url: fetchURLBySourceId,
+          success: (data) => {
+            if (data && data.length > 0) {
+              setUpChapterMarkers(data[0]);
+            }
+          },
+          dataType: 'jsonp'
+        });
+      }
+    },
+    dataType: 'jsonp'
+  });
```

Comment:
`$.ajax` is a type of promise! So should be able to do promise chaining here.

```suggestion
  $.ajax({ url: fetchUrlByBookId, dataType: 'jsonp' })
  .then(data => {
    if (data && data.length > 0) {
      return data;
    } else {
      // try sourceid
      return $.ajax({ url: `${baseURL}&source_records=ia:${this.bookId}`, dataType: 'jsonp' });
    }
  })
  .then(data => {
    if (data && data.length > 0) {
        setUpChapterMarkers(data[0]);
    }
  });
```